File: Common/RemoteStashServer.m
//
//  RemoteCopyClient.m
//  remotecopypaste
//
//  Created by Brice Rosenzweig on 14/05/2020.
//  Copyright © 2020 Brice Rosenzweig. All rights reserved.
//

#import "RemoteStashServer.h"
#include <arpa/inet.h>
#include <ifaddrs.h>
#if TARGET_IPHONE_SIMULATOR
#define TypeEN    "en1"
#else
#define TypeEN    "en0"
#endif


@import Criollo;

@interface RemoteStashServer ()
@property (nonatomic,retain) NSNetService * service;
@property (nonatomic,retain) GCDAsyncSocket * socket;
@property (nonatomic,retain) dispatch_queue_t worker;
@property (nonatomic,retain) CRHTTPServer * httpServer;
@property (nonatomic,assign) int port;
@end

@implementation RemoteStashServer

+(RemoteStashServer*)server:(NSObject<RemoteStashServerDelegate>*)delegate{
    RemoteStashServer * rv =[[RemoteStashServer alloc] init];
    if( rv ){
        rv.delegate = delegate;
        [rv startBroadCast];
    }
    return rv;
}

-(void)startHttpServer{
    self.httpServer = [[CRHTTPServer alloc] init];
    self.httpServer.isSecure = YES;
    
    self.httpServer.certificatePath = [NSBundle.mainBundle pathForResource:@"remotestash-ios-cert" ofType:@"pem"];
    self.httpServer.certificateKeyPath = [NSBundle.mainBundle pathForResource:@"remotestash-ios-key" ofType:@"pem"];

    [self.httpServer get:@"/status" block:^(CRRequest*req, CRResponse * res, CRRouteCompletionBlock next){
        NSData * data = [NSJSONSerialization dataWithJSONObject:@{@"items_count":@0} options:NSJSONWritingSortedKeys error:nil];
        if( data ){
            [res setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
            [res sendData:data];
        }
    }];
    [self.httpServer startListening:nil portNumber:self.port];
    if( [self getIPAddresses].count > 0){
        NSLog(@"https://%@:%@", [self getIPAddresses].firstObject, @(self.port));
    }
}

-(NSArray<NSString*>*)getIPAddresses{
    NSMutableArray * rv = [NSMutableArray array];
    
    char buffer[MAX(INET6_ADDRSTRLEN,INET_ADDRSTRLEN)];
    
    NSString *address = @"error";
    struct ifaddrs *interfaces = NULL;
    struct ifaddrs *temp_addr = NULL;
    int success = 0;
    // retrieve the current interfaces - returns 0 on success
    success = getifaddrs(&interfaces);
    if (success == 0) {
        // Loop through linked list of interfaces
        temp_addr = interfaces;
        while(temp_addr != NULL) {
            if(temp_addr->ifa_addr->sa_family == AF_INET || temp_addr->ifa_addr->sa_family == AF_INET6) {
                
                NSString * iname = [NSString stringWithUTF8String:temp_addr->ifa_name];
                
                // Check if interface is en0 which is the wifi connection on the iPhone
                if([iname isEqualToString:@TypeEN]) {
                    if (temp_addr->ifa_addr->sa_family == AF_INET) {
                        struct sockaddr_in *in_addr = (struct sockaddr_in*) temp_addr->ifa_addr;
                        inet_ntop(AF_INET, &in_addr->sin_addr, buffer, (socklen_t)sizeof(buffer));
                        address = [NSString stringWithUTF8String:buffer];
                        [rv addObject:address];
                    }/* ignore IPV6
                      else { // AF_INET6
                        struct sockaddr_in6 *in6_addr = (struct sockaddr_in6*) temp_addr->ifa_addr;
                        inet_ntop(AF_INET6, &in6_addr->sin6_addr, buffer, (socklen_t)sizeof(buffer));
                    }*/
                }
                
            }
            
            temp_addr = temp_addr->ifa_next;
        }
    }
    // Free memory
    freeifaddrs(interfaces);
    return rv;
}

-(void)startBroadCast{
    dispatch_queue_t queue = dispatch_queue_create("net.ro-z.worker", DISPATCH_QUEUE_SERIAL);
    self.worker = queue;

    // Find a free port number
    self.socket = [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:self.worker];
    [self.socket acceptOnPort:0 error:nil];
    self.port = self.socket.localPort;
    [self.socket disconnect];
    self.socket = nil;
    
    NSString * name = [NSString stringWithFormat:@"%@ RemoteStash", [[UIDevice currentDevice] name]];
    self.service = [[NSNetService alloc] initWithDomain:@"local." type:@"_remotestash._tcp" name:name port:self.port];
    [self.service publish];
    [self startHttpServer];
}

#pragma mark - NetService

-(void)netServiceDidPublish:(NSNetService *)sender{
}

#pragma mark - Socket

-(void)socket:(GCDAsyncSocket *)sock didAcceptNewSocket:(GCDAsyncSocket *)newSocket{
    
}

-(void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag{
}
@end

File: Sources/Common/RemoteStashServer.swift
//
//  RemoteStashServer.swift
//  remotestash
//
//  Created by Brice Rosenzweig on 30/01/2021.
//  Copyright © 2021 Brice Rosenzweig. All rights reserved.
//

import Foundation
import Criollo
import os

protocol RemoteStashServerDelegate {
    func serverStarted(_ server : RemoteStashServer)
    func server(_ server : RemoteStashServer, received : RemoteStashItem)
    func serverLastItem( _ server : RemoteStashServer) -> RemoteStashItem?
}

fileprivate let logger = Logger(subsystem: "net.ro-z.remotestash", category: "server")

class RemoteStashServer : NSObject,NetServiceDelegate,NetServiceBrowserDelegate {
    
    struct Status : Codable, CustomStringConvertible {
        enum CodingKeys : String, CodingKey {
            case itemsCount = "items_count"
            case last = "last"
        }
        
        var description: String{
            let last = self.last?.description ?? "none"
            return "Status(itemsCount: \(self.itemsCount), last: \(last))"
        }
        
        let itemsCount : Int
        let last : RemoteStashItem.Status?
    }
    
    let serverUUID :UUID
    let delegate : RemoteStashServerDelegate
    let name : String
    
    private let worker : DispatchQueue
    private var port : Int
    private var httpServer : CRHTTPServer? = nil
    private var service : NetService? = nil
    
    init(delegate : RemoteStashServerDelegate, name : String? = nil) {
        self.serverUUID = UUID()
        self.delegate = delegate
        self.name = name ?? "\(UIDevice.current.name) RemoteStash"
        self.worker = DispatchQueue(label: "net.ro-z.worker")
        self.port = 0
    }

    //MARK: - start/stop
    
    func start() {
        self.findPort()
        self.startBroadcast()
        self.startServer()
    }
    
    func stop(){
        self.service?.stop()
        self.httpServer?.stopListening()
    }

    //MARK: - get network info
    
    func findPort() {
        let port = AddressAndPort.availablePort()
        self.port = port?.port ?? 0
    }
    
    
    //MARK: start server and netservices
    
    func startServer() {
        let httpServer = CRHTTPServer()
        self.httpServer = httpServer
        
        httpServer.isSecure = true
        // Certificate created with
        // openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout remotestash-key.pem -out remotestash-cert.pem
        httpServer.certificatePath = Bundle.main.path(forResource: "remotestash-cert", ofType: "pem")
        httpServer.certificateKeyPath = Bundle.main.path(forResource: "remotestash-key", ofType: "pem")
        
        httpServer.get("/status") {
            (req, res, next) in
            let itemStatus : RemoteStashItem.Status? = self.delegate.serverLastItem(self)?.status
            let status : Status = Status(itemsCount: itemStatus != nil ? 1 : 0, last: itemStatus)
            if let data = try? JSONEncoder().encode(status) {
                res.addValue("application/json", forHTTPHeaderField: "Content-Type")
                res.send(data)
            }
        }
        
        httpServer.get("/pull"){
            (req,res,next) in
            if let item = self.delegate.serverLastItem(self) {
                item.prepare(request: req, into: res)
            }
        }
        
        httpServer.get("/last"){
            (req,res,next) in
            if let item = self.delegate.serverLastItem(self) {
                item.prepare(request: req, into: res)
            }
        }
        
        httpServer.post("/push"){
            (req,res,next) in
            let item = RemoteStashItem(request: req, response: res)
            self.delegate.server(self, received: item)
            res.send(["success":1])
            logger.info("pushed \(item)")
        }
        
        if httpServer.startListening(nil, portNumber: UInt(self.port)) {
            logger.info("started listening on \(self.port)")
        }else{
            logger.error("failed to started listening on \(self.port)")
        }
    }
    
    func startBroadcast() {
        let name = self.name
        let service = NetService(domain: "local.", type: "_remotestash._tcp", name: name, port: Int32(self.port))
        let txtRecord = [
            "temporary" : "yes".data(using: .utf8) ?? Data(),
            "uuid" : self.serverUUID.uuidString.data(using: .utf8) ?? Data()
        ]
        service.setTXTRecord(NetService.data(fromTXTRecord: txtRecord ) )
        
        logger.info("Starting broadcast \(self.port) \(self.serverUUID)")
        
        self.service = service
        self.service?.publish()
    }
    
    //MARK: - netService
    
    func netServiceDidPublish(_ sender: NetService) {
        logger.info("Published \(sender)")
    }
    
    func netServiceDidStop(_ sender: NetService) {
        logger.info("Stop \(sender)")
    }
    
}
